### PR TITLE
🌱 Add providerID index to get nodes

### LINKED ISSUE #4518: Use a providerID index to find a nodeRef for machines
**What steps did you take and what happened:**
We are looping over all nodes to find a nodeRef for machines.
https://github.com/kubernetes-sigs/cluster-api/blob/bfc6f80add5c21b8dc2b704951f42bc14708ebc4/controllers/machine_controller_noderef.go#L165

**What did you expect to happen:**
We should use an index based on the providerID,

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Cluster-api version:
- Minikube/KIND version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

/kind bug
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]

### PR DESCRIPTION
**Pull request by enxebre**
**What this PR does / why we need it**:
Add a providerID index to get nodes rather than requiring looping over all of them every time.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4518

### CODE REVIEW

**Comment by enxebre** [comment]
Test failing with 
```
    machine_controller_noderef_test.go:119: 
        Unexpected error:
            <*cache.ErrCacheNotStarted | 0x2b87348>: {}
            the cache is not started, can not read objects
```
I'm not being able to reproduce this locally.

File: controllers/noderefutil/providerid.go [hunk 1]
```diff
@@ -96,3 +97,11 @@ func (p *ProviderID) String() string {
 func (p *ProviderID) Validate() bool {
 	return p.CloudProvider() != "" && p.ID() != ""
 }
+
+// IndexKey returns a string concatenating the cloudProvider and the ID parts of the providerID.
+// E.g Format: cloudProvider://optional/segments/etc/id. Parsed: cloudProvider/id
+// This is useful to use the providerID as a reliable index between nodes and machines
+// as it guarantees the infra Providers contract.
+func (p *ProviderID) IndexKey() string {
+	return fmt.Sprintf("%s/%s", p.CloudProvider(), p.ID())
```
Comment (fabriziopandini) [inline controllers/noderefutil/providerid.go:106]: Wondering if dropping segments could lead to duplicated indexKey; e.g. looking at AWS or GCE, is it possible to have the same machine id in different regions? 
Would it be more straight forward/secure if we index for the entire providerID?

File: controllers/machine_controller_noderef_test.go [hunk 2]
```diff
@@ -20,106 +20,103 @@ import (
...
+	defer func(do ...client.Object) {
+		g.Expect(testEnv.Cleanup(ctx, do...)).To(Succeed())
+	}(ns, testCluster)
 
 	testCases := []struct {
-		name       string
-		providerID string
-		expected   *corev1.ObjectReference
-		err        error
+		node            *corev1.Node
```
Comment (fabriziopandini) [inline controllers/machine_controller_noderef_test.go:53]: Should we add name to describe the test case?

File: controllers/noderefutil/providerid.go (hunk 1 above)
Comment (enxebre) [reply controllers/noderefutil/providerid.go:106]: I was just trying to be consistent with existing assumptions, cloudProvider and ID is what's currently used to compare, see https://github.com/kubernetes-sigs/cluster-api/blob/master/controllers/noderefutil/providerid.go#L86

In a follow up I'd be in favour of changing the expectation for the cluster-api-providers to set exactly the same providerID the controller manager cloud-provider sets. 
We actually ensured this in AWS a while ago https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/1693.
Though there might be other provider specific reasons why this is not possible I'm not aware of.

Comment (enxebre) [reply controllers/noderefutil/providerid.go:106]: Created this to track the discussion https://github.com/kubernetes-sigs/cluster-api/issues/4526

File: controllers/machine_controller_noderef_test.go [hunk 3]
```diff
@@ -19,106 +19,128 @@ package controllers
 import (
 	"testing"
 
-	. "github.com/onsi/gomega"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	ctrl "sigs.k8s.io/controller-runtime"
```
Comment (vincepri) [inline controllers/machine_controller_noderef_test.go:26]: Could we group these with the other imports?

File: controllers/machine_controller_noderef.go [hunk 4]
```diff
@@ -163,30 +164,17 @@ func summarizeNodeConditions(node *corev1.Node) (corev1.ConditionStatus, string)
...
-				log.Error(err, "Failed to parse ProviderID", "node", node.Name)
-				continue
-			}
-
-			if providerID.Equals(nodeProviderID) {
-				return &node, nil
-			}
-		}
+	if err := c.List(ctx, &nodeList, client.MatchingFields{remote.NodeProviderIDIndex: providerID.IndexKey()}); err != nil {
+		return nil, err
```
Comment (vincepri) [inline controllers/machine_controller_noderef.go:169]: Just a note that field indexer don't work with the fake client (which we're trying not to use throughout CAPI, but we should make sure there are no tests with the fake client that use this.

File: controllers/machine_controller_noderef.go [hunk 5]
```diff
@@ -163,30 +164,17 @@ func summarizeNodeConditions(node *corev1.Node) (corev1.ConditionStatus, string)
...
+	}
+	if len(nodeList.Items) == 0 {
+		return nil, ErrNodeNotFound
+	}
 
-		if nodeList.Continue == "" {
-			break
-		}
+	if len(nodeList.Items) > 1 {
+		return nil, fmt.Errorf("unexpectedly found more than one Node matching the poviderID %s", providerID.String())
```
Comment (vincepri) [inline controllers/machine_controller_noderef.go:176]:
```suggestion
		return nil, fmt.Errorf("unexpectedly found more than one Node matching the providerID %s", providerID.String())
```

File: controllers/noderefutil/providerid.go (hunk 1 above)
Comment (vincepri) [reply controllers/noderefutil/providerid.go:106]: > Wondering if dropping segments could lead to duplicated indexKey; e.g. looking at AWS or GCE, is it possible to have the same machine id in different regions?

AWS or GCE machine names should be globally unique. cc @randomvariable @sedefsavas 

In general though, the ProviderID's id part is intended to be globally unique

File: controllers/remote/cluster_cache.go [hunk 6]
```diff
@@ -154,8 +159,15 @@ func (t *ClusterCacheTracker) newClusterAccessor(ctx context.Context, cluster cl
 		cancelFunc: cacheCtxCancel,
 	}
 
+	if err := cache.IndexField(ctx, &corev1.Node{}, NodeProviderIDIndex, indexNodeByProviderID); err != nil {
+		return nil, fmt.Errorf("failed to index field %s: %w", NodeProviderIDIndex, err)
+	}
```
Comment (vincepri) [inline controllers/remote/cluster_cache.go:164]: Should we create this indexer for every client?

File: controllers/machine_controller_noderef_test.go [hunk 7]
```diff
@@ -19,106 +19,128 @@ package controllers
...
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/client-go/kubernetes/scheme"
-	"k8s.io/client-go/tools/record"
 	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
 	"sigs.k8s.io/cluster-api/controllers/noderefutil"
+	"sigs.k8s.io/cluster-api/controllers/remote"
+	"sigs.k8s.io/cluster-api/util"
 	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/log"
```
Comment (vincepri) [inline controllers/machine_controller_noderef_test.go:35]: Do we need to use the global logger here?

**Review by vincepri — CHANGES_REQUESTED** [review]

File: controllers/remote/cluster_cache.go [hunk 8]
```diff
@@ -154,8 +159,15 @@ func (t *ClusterCacheTracker) newClusterAccessor(ctx context.Context, cluster cl
...
+
 	// Start the cache!!!
 	go cache.Start(cacheCtx)
+	if !cache.WaitForCacheSync(ctx) {
+		return nil, fmt.Errorf("failed waiting for cache for remote cluster %v to sync: %w", cluster, err)
+	}
```
Comment (vincepri) [inline controllers/remote/cluster_cache.go:170]: This is a good improvement, although could we split it out in a different PR to have its own release note? It might also be worth to backport.

File: controllers/machine_controller_noderef_test.go (hunk 7 above)
Comment (enxebre) [reply controllers/machine_controller_noderef_test.go:35]: It's being used to pass a NullLogger here https://github.com/kubernetes-sigs/cluster-api/pull/4521/files#diff-7b35b89626a4594a7d7296ab37621dd6998ee1068830357e042c12f00967d0c1R110

Happy to change if there's a more reasonable way.

File: controllers/remote/cluster_cache.go (hunk 8 above)
Comment (enxebre) [reply controllers/remote/cluster_cache.go:170]: I think we need it for the test to not go flaky. How about having its own commit on this PR?

File: controllers/remote/cluster_cache.go (hunk 6 above)
Comment (enxebre) [reply controllers/remote/cluster_cache.go:164]: It's being created for each newClusterAccessor, which others are you thinking of?

**Comment by enxebre** [comment]
```
sigs.k8s.io/cluster-api/controlplane/kubeadm/controllers: TestReconcileUnhealthyMachines/Remediation_does_not_happen_if_number_of_machines_lower_than_desired expand_more0s | sigs.k8s.io/cluster-api/controlplane/kubeadm/controllers: TestReconcileUnhealthyMachines/Remediation_does_not_happen_if_number_of_machines_lower_than_desired expand_more | 0s
-- | -- | --
sigs.k8s.io/cluster-api/controlplane/kubeadm/controllers: TestReconcileUnhealthyMachines/Remediation_does_not_happen_if_number_of_machines_lower_than_desired expand_more | 0s
sigs.k8s.io/cluster-api/controlplane/kubeadm/controllers: TestReconcileUnhealthyMachines expand_more1s | sigs.k8s.io/cluster-api/controlplane/kubeadm/controllers: TestReconcileUnhealthyMachines expand_more | 1s
sigs.k8s.io/cluster-api/controlplane/kubeadm/controllers: TestReconcileUnhealthyMachines expand_more | 1s
sigs.k8s.io/cluster-api/exp/addons/controllers: TestAPIs expand_more | sigs.k8s.io/cluster-api/exp/addons/controllers: TestAPIs expand_more
sigs.k8s.io/cluster-api/exp/addons/controllers: TestAPIs expand_more
```

/test pull-cluster-api-test-main

File: controllers/noderefutil/providerid.go (hunk 1 above)
Comment (randomvariable) [reply controllers/noderefutil/providerid.go:106]: Instance ID may not be unique across regions, but the provider ID does encode the AZ, so this should be fine.

Comment (enxebre) [reply controllers/noderefutil/providerid.go:106]: >Instance ID may not be unique across regions, but the provider ID does encode the AZ (e.g. eu-west-1a), so this should be fine.

That invalidates our equality check.
Please let's contain this discussion in this thread to avoid having it scattered https://github.com/kubernetes-sigs/cluster-api/issues/4526#issuecomment-827473195

File: controllers/remote/cluster_cache.go (hunk 6 above)
Comment (vincepri) [reply controllers/remote/cluster_cache.go:164]: Given that this is a remote client, we're making an assumption here that we'd get nodes and we want to index them

File: controllers/machine_controller_noderef.go [hunk 9]
```diff
@@ -163,30 +164,17 @@ func summarizeNodeConditions(node *corev1.Node) (corev1.ConditionStatus, string)
...
+	}
+	if len(nodeList.Items) == 0 {
+		return nil, ErrNodeNotFound
+	}
 
-		if nodeList.Continue == "" {
-			break
-		}
+	if len(nodeList.Items) > 1 {
+		return nil, fmt.Errorf("unexpectedly found more than one Node matching the providerID %s", providerID.String())
```
Comment (vincepri) [inline controllers/machine_controller_noderef.go:176]: Instead of erroring out, could we still use the for loop and validate the equality of the provider id?

We should also make to fallback to list all nodes in case the list doesn't return any items. This could save us later in case we need changes to how the providerIDs are formed.

Comment (enxebre) [reply controllers/machine_controller_noderef.go:176]: If we are using the same indexKey we use for the equality check i.e `cloud/id`, there's no additional value on doing the two loops you describe above right? the client indexing already had to do that loop underneath.

File: controllers/remote/cluster_cache.go (hunk 6 above)
Comment (enxebre) [reply controllers/remote/cluster_cache.go:164]: It seems reasonable to me, if you disagree we could move the indexing logic to the Watch method.

File: controllers/noderefutil/providerid.go (hunk 1 above)
Comment (enxebre) [reply controllers/noderefutil/providerid.go:106]: I'll resolve this so we keep the discussion in thread.

File: controllers/machine_controller_noderef_test.go [hunk 10]
```diff
@@ -20,105 +20,125 @@ import (
...
-			node, err := r.getNode(ctx, client, providerID)
-			if test.err == nil {
-				g.Expect(err).To(BeNil())
-			} else {
-				gt.Expect(err).NotTo(BeNil())
-				gt.Expect(err).To(Equal(test.err), "Expected error %v, got %v", test.err, err)
-			}
+	nodesToCleanup := make([]client.Object, 0, len(testCases))
+	for _, tc := range testCases {
+		g.Expect(testEnv.Create(ctx, tc.node)).To(BeNil())
```
Comment (sbueringer) [inline controllers/machine_controller_noderef_test.go:100]: Just out of curiosity (and I'm new to most of this, so I could be wrong about anything I assume :)). 

We're using the same testEnv for multiple tests (in parallel?). Usually the tests are separated from each other because we deploy resources into a test-specific namespace. Could it lead to strange side effects in other tests when we deploy cluster-wide resources like Nodes?

Comment (enxebre) [reply controllers/machine_controller_noderef_test.go:100]: Sounds reasonable to me, I renamed the nodes to be test-name-prefixed as a good practice. Thanks!

**Comment by sbueringer** [comment]
As far as I could figure out the implications +> lgtm

File: controllers/machine_controller_noderef.go (hunk 9 above)
Comment (vincepri) [reply controllers/machine_controller_noderef.go:176]: If for whatever reason the index isn't registered or available, or the fake client is used this might fail

Comment (enxebre) [reply controllers/machine_controller_noderef.go:196]: ok I updated to fallback and loop over the whole list when no nodes are found via index.

I'd error for the case where more than 1 are found rather than overcomplicating the func by reimplementing the logic to count with no indexes. At the moment for this case we are just returning the first one, erroring seems an improvement. Please let me know if you disagree.

File: controllers/remote/cluster_cache.go (hunk 6 above)
Comment (vincepri) [reply controllers/remote/cluster_cache.go:164]: We should move this under the `Watch` method maybe? We should only ever create this index if a `&corev1.Node{}` is passed in the WatchInput struct as `Kind` object.

**Review by vincepri — CHANGES_REQUESTED** [review]

File: controllers/remote/cluster_cache.go [hunk 11]
```diff
@@ -247,6 +273,18 @@ func (t *ClusterCacheTracker) Watch(ctx context.Context, input WatchInput) error
 		return nil
 	}
 
+	_, ok := input.Kind.(*corev1.Node)
+	if ok {
+		a.cache.Stop()
```
Comment (vincepri) [inline controllers/remote/cluster_cache.go:278]: Is this required?

Comment (enxebre) [reply controllers/remote/cluster_cache.go:278]: without it, test was complaining about informer already starter. But It's still failing, I need to dig into it a bit.

Comment (enxebre) [reply controllers/remote/cluster_cache.go]: @vincepri I changed the approach slightly. I added a field `Indexes` to `WatchInput` that is passed through `getClusterAccessorLH`.
So now is up to the outer function e.g `watchNodes` to define their opinionated indexes.

Comment (enxebre) [reply controllers/remote/cluster_cache.go]: heh this makes `TestWatches` fail because `WatchClusterNodes` is gated by `ControlPlaneInitializedCondition`, which depends on the control machine having a nodeRef, so when `getNode` is called for that Machine the the watcher is still not running nor the index initialised so it fails with `"Failed to retrieve Node by ProviderID" "error"="Index with name field:spec.providerID does not exist`

Comment (enxebre) [reply controllers/remote/cluster_cache.go]: @vincepri Given the above I'm starting to think that getting back to set the index in `newClusterAccessor` unconditionally might be the most reasonable thing to do. Thoughts?

Comment (vincepri) [reply controllers/remote/cluster_cache.go]: Wouldn't `TestWatches` create a new remote cache with an index?

File: controllers/machine_controller.go [hunk 12]
```diff
@@ -645,6 +666,13 @@ func (r *MachineReconciler) watchClusterNodes(ctx context.Context, cluster *clus
 		Watcher:      r.controller,
 		Kind:         &corev1.Node{},
 		EventHandler: handler.EnqueueRequestsFromMapFunc(r.nodeToMachine),
+		Indexes: []remote.Index{
+			{
+				Obj:          &corev1.Node{},
+				Field:        NodeProviderIDIndex,
+				ExtractValue: indexNodeByProviderID,
+			},
+		},
```
Comment (vincepri) [inline controllers/machine_controller.go:675]: I don't think this should be defined in here, but rather we should expose some options to `NewClusterCacheTracker` to add indexes when the ClusterTracker is initialized. 

Only indexes defined at the top level will be able to be used throughout the lifecycle of a ClusterCacheTracker, and if a user tries to access something an unknown index, it's up to them to define them in the list of indexes.

**Review by vincepri — CHANGES_REQUESTED** [review]

File: controllers/remote/cluster_cache.go (hunk 11 above)
Comment (enxebre) [reply controllers/remote/cluster_cache.go]: the way it is right now it fails because the code is broken because getNodes tries to use and index which doesn't exist in the client it uses https://github.com/kubernetes-sigs/cluster-api/blob/1f90fa0dc94a66e232cd6a67af01ad3a23018e6b/controllers/machine_controller_noderef.go#L56 so we'd need to let `GetClient` to receive a list on indexes.

File: controllers/machine_controller.go (hunk 12 above)
Comment (enxebre) [reply controllers/machine_controller.go:675]: that sounds good to me, though it would mean all clusterAccessor/clients for a tracker need to use the same indexes defeating the original attempt indexing only each client for its watched kind https://github.com/kubernetes-sigs/cluster-api/pull/4521#discussion_r620642643
Unless we store some relation kind/index in `NewClusterCacheTracker` and then we use that value to discriminate and pass only the right indexes through `getClusterAccessorLH` from func `Watch`

Comment (enxebre) [reply controllers/machine_controller.go]: @vincepri updated PTAL.

Comment (vincepri) [reply controllers/machine_controller.go]: Why should we make the assumption that this client is always going to be used to index nodes?

Comment (enxebre) [reply controllers/machine_controller.go]: Seemed convenient to me given I don't think of `remote` as a general purpose library but rather as something we are the main consumers and we know we need to index nodes. I'll update to drop the default and set the index when instantiating the tracker.

Comment (enxebre) [reply controllers/machine_controller.go]: Updated.

File: api/v1alpha4/common_types.go [hunk 13]
```diff
@@ -102,6 +102,10 @@ var (
 const (
 	// MachineNodeNameIndex is used by the Machine Controller to index Machines by Node name, and add a watch on Nodes.
 	MachineNodeNameIndex = "status.nodeRef.name"
+
+	// NodeProviderIDIndex is used to index Nodes by ProviderID. Remote caches use this to find Nodes in a workload cluster
+	// out of management cluster machine.spec.providerID.
+	NodeProviderIDIndex = "spec.providerID"
```
Comment (vincepri) [inline api/v1alpha4/common_types.go:108]: Should this constant be under `norerefutil` package?

**Comment by vincepri** [comment]
/assign @CecileRobertMichon 
for another set of 👀

**Comment by CecileRobertMichon** [comment]
I completely missed this, @enxebre are you still waiting for my review?

File: api/v1alpha4/common_types.go (hunk 13 above)
Comment (vincepri) [reply api/v1alpha4/common_types.go:108]: Comment still stands

File: controllers/machine_controller_noderef.go [hunk 14]
```diff
@@ -170,29 +169,41 @@ func summarizeNodeConditions(node *corev1.Node) (corev1.ConditionStatus, string)
...
 			}
 		}
 
-		if nodeList.Continue == "" {
-			break
-		}
+		return nil, ErrNodeNotFound
+	}
+
+	if len(nodeList.Items) > 1 {
```
Comment (vincepri) [inline controllers/machine_controller_noderef.go:204]: Can't `nodeList` still be zero at this point?

File: controlplane/kubeadm/internal/cluster_test.go [hunk 15]
```diff
@@ -27,6 +27,8 @@ import (
 	"testing"
 	"time"
 
+	"sigs.k8s.io/cluster-api/controllers/noderefutil"
+
```
Comment (vincepri) [inline controlplane/kubeadm/internal/cluster_test.go:31]:
```suggestion
```

File: controllers/remote/cluster_cache_tracker_test.go [hunk 16]
```diff
@@ -20,6 +20,8 @@ import (
 	"context"
 	"testing"
 
+	"sigs.k8s.io/cluster-api/controllers/noderefutil"
+
```
Comment (vincepri) [inline controllers/remote/cluster_cache_tracker_test.go:24]:
```suggestion
```

File: controllers/remote/cluster_cache.go [hunk 17]
```diff
@@ -235,6 +248,13 @@ type Watcher interface {
 	Watch(src source.Source, eventHandler handler.EventHandler, predicates ...predicate.Predicate) error
 }
 
+// Index is a helper to model the info passed to cache.IndexField.
+type Index struct {
+	Obj          client.Object
```
Comment (vincepri) [inline controllers/remote/cluster_cache.go:253]:
```suggestion
	Object          client.Object
```

File: controllers/machine_controller_noderef.go (hunk 14 above)
Comment (CecileRobertMichon) [reply controllers/machine_controller_noderef.go:204]: it shouldn't be, if it was 0 before and we entered the for loop we should have returned either the node or `ErrNodeNotFound` if it wasn't found

Comment (CecileRobertMichon) [reply controllers/machine_controller_noderef.go:204]: for safety this can be 

```suggestion
	if len(nodeList.Items) != 1 {
```

though

File: controllers/noderefutil/providerid.go [hunk 18]
```diff
@@ -100,3 +101,11 @@ func (p *ProviderID) String() string {
 func (p *ProviderID) Validate() bool {
 	return p.CloudProvider() != "" && p.ID() != ""
 }
+
+// IndexKey returns a string concatenating the cloudProvider and the ID parts of the providerID.
+// E.g Format: cloudProvider://optional/segments/etc/id. IndexKey: cloudProvider/id
+// This is useful to use the providerID as a reliable index between nodes and machines
+// as it guarantees the infra Providers contract.
+func (p *ProviderID) IndexKey() string {
+	return fmt.Sprintf("%s/%s", p.CloudProvider(), p.ID())
```
Comment (CecileRobertMichon) [inline controllers/noderefutil/providerid.go:110]: I know there was a previous discussion on this but looks like it wasn't resolved, wouldn't this break with some providers? Does this also break for Azure when the vmss ID can be the same across different scale sets @alexeldeib ?

Comment (alexeldeib) [reply controllers/noderefutil/providerid.go:110]: Its already broken for VMSS, this is no more broken than the existing logic unfortunately.

I think this PR is good, it doesn’t affect the existing behavior much besides adding the index and falling back to listing. we really need to do full provider ID comparisons everywhere for a fix on the issue you mentioned.

File: controllers/machine_controller_noderef.go (hunk 14 above)
Comment (enxebre) [reply controllers/machine_controller_noderef.go]: Added a test.

File: controllers/noderefutil/providerid.go (hunk 18 above)
Comment (enxebre) [reply controllers/noderefutil/providerid.go:110]: tracked here https://github.com/kubernetes-sigs/cluster-api/issues/4526

**Comment by enxebre** [comment]
PTAL @vincepri @CecileRobertMichon

**Review by vincepri — APPROVED** [review]
/lgtm

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
/approve
